File: java/src/epp/Lexer.java
package epp;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class Lexer {
  public static final class Result {
    public final List<Token> tokens;
    public final List<LexError> errors;
    public Result(List<Token> tokens, List<LexError> errors) {
      this.tokens = tokens;
      this.errors = errors;
    }
  }

  private static final Map<String, TokenType> KEYWORDS = new HashMap<>();
  static {
    KEYWORDS.put("set", TokenType.SET);
    KEYWORDS.put("to", TokenType.TO);
    KEYWORDS.put("say", TokenType.SAY);
    KEYWORDS.put("ask", TokenType.ASK);
    KEYWORDS.put("into", TokenType.INTO);
    KEYWORDS.put("if", TokenType.IF);
    KEYWORDS.put("then", TokenType.THEN);
    KEYWORDS.put("else", TokenType.ELSE);
    KEYWORDS.put("end", TokenType.END);
    KEYWORDS.put("while", TokenType.WHILE);
    KEYWORDS.put("do", TokenType.DO);
    KEYWORDS.put("try", TokenType.TRY);
    KEYWORDS.put("catch", TokenType.CATCH);
    KEYWORDS.put("function", TokenType.FUNCTION);
    KEYWORDS.put("return", TokenType.RETURN);
    KEYWORDS.put("and", TokenType.AND);
    KEYWORDS.put("at", TokenType.AT);
    KEYWORDS.put("is", TokenType.IS);
    KEYWORDS.put("not", TokenType.NOT);
    KEYWORDS.put("equal", TokenType.EQUAL);
    KEYWORDS.put("greater", TokenType.GREATER);
    KEYWORDS.put("less", TokenType.LESS);
    KEYWORDS.put("than", TokenType.THAN);
    KEYWORDS.put("true", TokenType.TRUE);
    KEYWORDS.put("false", TokenType.FALSE);
  }

  private final String text;
  private int i = 0;
  private int line = 1;
  private int col = 1;

  private final List<Token> tokens = new ArrayList<>();
  private final List<LexError> errors = new ArrayList<>();

  private Lexer(String text) { this.text = text; }

  public static Result lex(String text) {
    return new Lexer(text).run();
  }

  private Result run() {
    while (!atEnd()) {
      char c = peek(0);
      Span sp = spanNow();

      if (c == ' ' || c == '\t' || c == '\r') { advance(); continue; }
      if (c == '#') { while (!atEnd() && peek(0) != '\n') advance(); continue; }
      if (c == '\n') { advance(); add(TokenType.NEWLINE, "\n", sp); continue; }

      switch (c) {
        case '[': advance(); add(TokenType.LBRACKET, "[", sp); continue;
        case ']': advance(); add(TokenType.RBRACKET, "]", sp); continue;
        case '(': advance(); add(TokenType.LPAREN, "(", sp); continue;
        case ')': advance(); add(TokenType.RPAREN, ")", sp); continue;
        case ',': advance(); add(TokenType.COMMA, ",", sp); continue;
        case '+': advance(); add(TokenType.PLUS, "+", sp); continue;
        case '-': advance(); add(TokenType.MINUS, "-", sp); continue;
        case '*': advance(); add(TokenType.STAR, "*", sp); continue;
        case '/': advance(); add(TokenType.SLASH, "/", sp); continue;
        case '"': readString(sp); continue;
        default: break;
      }

      if (isDigit(c)) { readNumber(sp); continue; }
      if (isIdentStart(c)) { readIdent(sp); continue; }

      errors.add(new LexError("Unexpected character: '" + c + "'", sp));
      advance();
    }

    tokens.add(new Token(TokenType.EOF, "", new Span(line, col)));
    return new Result(tokens, errors);
  }

  private void readString(Span sp) {
    advance(); // opening quote
    StringBuilder sb = new StringBuilder();
    while (!atEnd() && peek(0) != '"') {
      char ch = advance();
      if (ch == '\\' && !atEnd()) {
        char esc = advance();
        switch (esc) {
          case 'n': sb.append('\n'); break;
          case 't': sb.append('\t'); break;
          case '"': sb.append('"'); break;
          case '\\': sb.append('\\'); break;
          default: sb.append(esc); break;
        }
      } else {
        sb.append(ch);
      }
    }
    if (atEnd() || peek(0) != '"') {
      errors.add(new LexError("Unterminated string literal", sp));
      return;
    }
    advance(); // closing quote
    add(TokenType.STRING, sb.toString(), sp);
  }

  private void readNumber(Span sp) {
    StringBuilder sb = new StringBuilder();
    while (!atEnd() && isDigit(peek(0))) sb.append(advance());
    if (!atEnd() && peek(0) == '.' && !atEnd(i + 1) && isDigit(peek(1))) {
      sb.append(advance()); // '.'
      while (!atEnd() && isDigit(peek(0))) sb.append(advance());
    }
    add(TokenType.NUMBER, sb.toString(), sp);
  }

  private void readIdent(Span sp) {
    StringBuilder sb = new StringBuilder();
    while (!atEnd() && isIdentChar(peek(0))) sb.append(advance());
    String raw = sb.toString();
    String lowered = raw.toLowerCase();
    TokenType kw = KEYWORDS.get(lowered);
    if (kw != null) add(kw, lowered, sp);
    else add(TokenType.IDENT, raw, sp);
  }

  private void add(TokenType t, String lexeme, Span sp) {
    tokens.add(new Token(t, lexeme, sp));
  }

  private Span spanNow() { return new Span(line, col); }

  private boolean atEnd() { return i >= text.length(); }
  private boolean atEnd(int idx) { return idx >= text.length(); }

  private char peek(int k) {
    int idx = i + k;
    if (idx >= text.length()) return '\0';
    return text.charAt(idx);
  }

  private char advance() {
    char c = text.charAt(i++);
    if (c == '\n') { line++; col = 1; }
    else col++;
    return c;
  }

  private static boolean isDigit(char c) { return c >= '0' && c <= '9'; }
  private static boolean isIdentStart(char c) { return Character.isLetter(c) || c == '_'; }
  private static boolean isIdentChar(char c) { return Character.isLetterOrDigit(c) || c == '_'; }
}

File: src/lexer.cpp
#include "lexer.h"

#include <cctype>
#include <unordered_map>

namespace epp {

static bool isIdentStart(char c) {
  return std::isalpha(static_cast<unsigned char>(c)) || c == '_';
}

static bool isIdentChar(char c) {
  return std::isalnum(static_cast<unsigned char>(c)) || c == '_';
}

static std::string lowerAscii(std::string s) {
  for (auto& ch : s) ch = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
  return s;
}

LexResult lex(const Source& src) {
  static const std::unordered_map<std::string, TokenType> kKeywords = {
    {"set", TokenType::Set},
    {"to", TokenType::To},
    {"say", TokenType::Say},
    {"ask", TokenType::Ask},
    {"into", TokenType::Into},
    {"if", TokenType::If},
    {"then", TokenType::Then},
    {"else", TokenType::Else},
    {"end", TokenType::End},
    {"while", TokenType::While},
    {"do", TokenType::Do},
    {"try", TokenType::Try},
    {"catch", TokenType::Catch},
    {"function", TokenType::Function},
    {"return", TokenType::Return},
    {"and", TokenType::And},
    {"or", TokenType::Or},
    {"at", TokenType::At},
    {"is", TokenType::Is},
    {"not", TokenType::Not},
    {"equal", TokenType::Equal},
    {"greater", TokenType::Greater},
    {"less", TokenType::Less},
    {"than", TokenType::Than},
    {"true", TokenType::True},
    {"false", TokenType::False},
  };

  LexResult out;
  const std::string& s = src.text;
  size_t i = 0;
  int line = 1;
  int col = 1;

  auto spanNow = [&]() -> Span { return Span{line, col}; };

  auto advance = [&]() -> char {
    char c = s[i++];
    if (c == '\n') {
      line++;
      col = 1;
    } else {
      col++;
    }
    return c;
  };

  auto peek = [&](size_t k = 0) -> char {
    if (i + k >= s.size()) return '\0';
    return s[i + k];
  };

  auto add = [&](TokenType type, std::string lexeme, Span sp) {
    out.tokens.push_back(Token{type, std::move(lexeme), sp});
  };

  while (i < s.size()) {
    char c = peek();
    Span sp = spanNow();

    if (c == ' ' || c == '\t' || c == '\r') {
      advance();
      continue;
    }

    if (c == '#') {
      while (i < s.size() && peek() != '\n') advance();
      continue;
    }

    if (c == '/' && peek(1) == '/') {
      while (i < s.size() && peek() != '\n') advance();
      continue;
    }

    if (c == '\n') {
      advance();
      add(TokenType::Newline, "\n", sp);
      continue;
    }

    switch (c) {
      case '[': advance(); add(TokenType::LBracket, "[", sp); continue;
      case ']': advance(); add(TokenType::RBracket, "]", sp); continue;
      case '(': advance(); add(TokenType::LParen, "(", sp); continue;
      case ')': advance(); add(TokenType::RParen, ")", sp); continue;
      case ',': advance(); add(TokenType::Comma, ",", sp); continue;
      case '+': advance(); add(TokenType::Plus, "+", sp); continue;
      case '-': advance(); add(TokenType::Minus, "-", sp); continue;
      case '*': advance(); add(TokenType::Star, "*", sp); continue;
      case '/': advance(); add(TokenType::Slash, "/", sp); continue;
      case '"': {
        advance(); // opening quote
        std::string str;
        while (i < s.size() && peek() != '"') {
          char ch = advance();
          if (ch == '\\' && i < s.size()) {
            char esc = advance();
            switch (esc) {
              case 'n': str.push_back('\n'); break;
              case 't': str.push_back('\t'); break;
              case '"': str.push_back('"'); break;
              case '\\': str.push_back('\\'); break;
              default: str.push_back(esc); break;
            }
          } else {
            str.push_back(ch);
          }
        }
        if (peek() != '"') {
          out.errors.push_back(LexError{"Unterminated string literal", sp});
          break;
        }
        advance(); // closing quote
        add(TokenType::String, str, sp);
        continue;
      }
      default: break;
    }

    if (std::isdigit(static_cast<unsigned char>(c))) {
      std::string num;
      while (std::isdigit(static_cast<unsigned char>(peek()))) num.push_back(advance());
      if (peek() == '.' && std::isdigit(static_cast<unsigned char>(peek(1)))) {
        num.push_back(advance());
        while (std::isdigit(static_cast<unsigned char>(peek()))) num.push_back(advance());
      }
      add(TokenType::Number, num, sp);
      continue;
    }

    if (isIdentStart(c)) {
      std::string id;
      while (isIdentChar(peek())) id.push_back(advance());
      std::string lowered = lowerAscii(id);
      auto it = kKeywords.find(lowered);
      if (it != kKeywords.end()) {
        add(it->second, lowered, sp);
      } else {
        add(TokenType::Identifier, id, sp);
      }
      continue;
    }

    out.errors.push_back(LexError{std::string("Unexpected character: '") + c + "'", sp});
    advance();
  }

  out.tokens.push_back(Token{TokenType::Eof, "", Span{line, col}});
  return out;
}

} // namespace epp

